### PR TITLE
New class for triangle

### PR DESCRIPTION
**Pull request by lbartoletti**
Hello,

I propose this new class which comes in part from my old plugin (CADDigitize). Almost all the code comes from this one.
I think, that this class is a first step for the OGC specification.

There are still tests and additions to make, but I prefer to propose it to you now to have your return.

Regards

### CODE REVIEW

File: src/core/geometry/qgswkbtypes.h [hunk 1]
```diff
@@ -114,9 +117,11 @@ class CORE_EXPORT QgsWkbTypes
       CurvePolygonZM = 3010,
       MultiCurveZM = 3011,
       MultiSurfaceZM = 3012,
+      TriangleZM = 3017,
       Point25D = 0x80000001,
       LineString25D,
       Polygon25D,
+      Triangle25D,
```
Comment (rouault) [inline src/core/geometry/qgswkbtypes.h]: Adding Triangle25D here will break the values of later enumeration items. xxx25D are for legacy stuff. No need for that for newer geometry types

**Review by rouault — CHANGES_REQUESTED** [review]

File: src/core/geometry/qgsgeometryutils.h [hunk 2]
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
      */
     static QgsPointV2 midpoint( const QgsPointV2& pt1, const QgsPointV2& pt2 );
 
+    /** Return the slope of a line defined by points pt1 and pt2.
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.h]: Small style change, but we're moving away from explicit listing of all params in cases like this (where the meaning has already been described). It's an nice touch to change "defined by points P1 and p2" to "defined by points \a p1 and \a p2" so that doxygen formats these as b arguments.

File: src/core/geometry/qgsgeometryutils.h [hunk 3]
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
...
+     * @param pt1 first point.
+     * @param pt2 second point.
+     * @return The slope of this linear entity, or infinity if vertical
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.h]: Probably worth mentioning that the slope returned is a gradient, and not an angle

Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.h]: Actually - maybe better to rename this method as "gradient" for clarity.

File: src/core/geometry/qgsgeometryutils.h [hunk 4]
```diff
@@ -291,6 +291,31 @@ class CORE_EXPORT QgsGeometryUtils
...
+     */
+    static double slope( const QgsPointV2& pt1, const QgsPointV2& pt2 );
+
+    /** Return the coefficients (a, b, c for equation "ax + by + c = 0") of a line defined by points pt1 and pt2.
+     * @param pt1 first point.
+     * @param pt2 second point.
+     * @return The coefficients of this linear entity
+     * @note added in QGIS 3.0
+     */
+    static QVector<double> coefficients( const QgsPointV2& pt1, const QgsPointV2& pt2 );
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.h]: Instead of returning a vector I think a/b/c should be arguments by reference

File: python/core/geometry/qgsgeometryutils.sip [hunk 5]
```diff
@@ -67,4 +67,10 @@ class QgsGeometryUtils
 
     static QgsPointV2 midpoint (const QgsPointV2& pt1, const QgsPointV2& pt2);
 
+    static double slope( const QgsPointV2& pt1, const QgsPointV2& pt2 );
+
+    static QVector<double> coefficients(const QgsPointV2 &pt1, const QgsPointV2 &pt2);
+
+    static QgsLineString* perpendicularSegment( const QgsPointV2& p, const QgsPointV2& s1, const QgsPointV2& s2 );
```
Comment (nyalldawson) [inline python/core/geometry/qgsgeometryutils.sip]: Needs a "/Factory/" annotation

File: src/core/geometry/qgsgeometryutils.cpp [hunk 6]
```diff
@@ -834,6 +834,82 @@ QgsPointV2 QgsGeometryUtils::midpoint( const QgsPointV2 &pt1, const QgsPointV2 &
...
+    coef.append( pt2.x() - pt1.x() );
+    coef.append( pt1.x() * pt2.y() - pt1.y() * pt2.x() );
+  }
+
+  return coef;
+}
+
+QgsLineString* QgsGeometryUtils::perpendicularSegment( const QgsPointV2 &p, const QgsPointV2 &s1, const QgsPointV2 &s2 )
+{
+  QgsLineString *line = new QgsLineString;
```
Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.cpp]: I'd like to see us move toward the more modern c++11 approach of using a unique_ptr here, which is only released when the object is returned

File: src/core/geometry/qgstriangle.cpp [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+  clear();
+
+  QPair<QgsWkbTypes::Type, QString> parts = QgsGeometryUtils::wktReadBlock( wkt );
+
+  if ( QgsWkbTypes::geometryType( parts.first ) != QgsWkbTypes::PolygonGeometry )
+    return false;
+
+  mWkbType = parts.first;
+
+  QString defaultChildWkbType = QStringLiteral( "LineString%1%2" ).arg( is3D() ? "Z" : "", isMeasure() ? "M" : "" );
```
Comment (nyalldawson) [inline src/core/geometry/qgstriangle.cpp]: Use QString() instead of "" here

**Review by nyalldawson — COMMENTED** [review]
Looking great so far!

File: src/core/geometry/qgsgeometryutils.h (hunk 4 above)
Comment (lbartoletti) [inline src/core/geometry/qgsgeometryutils.h]: OK. Do you want the same correction for:
QVector<double> lengths() const;
QVector<double> angles() const;
QVector<QgsLineString *> altitudes( ) const;
QVector<QgsLineString *> medians( ) const; 
QVector<QgsLineString *> bisectors( double lengthTolerance = 0.0001 ) const;

Comment (nyalldawson) [inline src/core/geometry/qgsgeometryutils.h]: No, I think those are fine to return as lists since the order isn't important

**Comment by nyalldawson** [comment]
Looking great to me! Do you have a to-do list for remaining changes before this is merge ready?

**Comment by lbartoletti** [comment]
My TODO:
- Add python binding
- Add triangle for QgsExpression
- Write full (best) doc
- Add some tests
- Check constructor with 3 points (no double point, colinearity). Maybe other constructors (QgsPoint, QPointF)

I'll do it at the weekend and you can merge it if it's ok.

Future TODO:
- export circumscribed and inscribed circle to QgsCircle (new class. I will work on it)

**Comment by nyalldawson** [comment]
Do you still get a crash if you change altitudes, medians and bisectors to return QVector< QgsLineString > (ie not pointers)? I think this might be a safer choice in any case.

File: src/core/qgsexpression.cpp [hunk 8]
```diff
@@ -3991,6 +4024,7 @@ const QList<QgsExpression::Function*>& QgsExpression::Functions()
     << new StaticFunction( QStringLiteral( "make_point_m" ), 3, fcnMakePointM, QStringLiteral( "GeometryGroup" ) )
     << new StaticFunction( QStringLiteral( "make_line" ), -1, fcnMakeLine, QStringLiteral( "GeometryGroup" ) )
     << new StaticFunction( QStringLiteral( "make_polygon" ), -1, fcnMakePolygon, QStringLiteral( "GeometryGroup" ) )
+    << new StaticFunction( QStringLiteral( "make_triangle" ), -1, fcnMakeTriangle, QStringLiteral( "GeometryGroup" ) )
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: Can you change this to use the new style which takes a ParameterList? That will correctly handle detection of invalid arguments.

File: src/core/qgsexpression.cpp [hunk 9]
```diff
@@ -2187,6 +2188,38 @@ static QVariant fcnMakePolygon( const QVariantList& values, const QgsExpressionC
...
+
+  QgsTriangle *tr = new QgsTriangle();
+  QgsLineString* lineString = new QgsLineString();
+  lineString->clear();
+
+  Q_FOREACH ( const QVariant& value, values )
+  {
+    QgsGeometry geom = getGeometry( value, parent );
+    if ( geom.isNull() )
+      continue;
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: What happens in the case that any of these continues are hit? Won't you be setting the exterior ring to a linestring with < 3 vertices?

I think it's better to flag an evaluation error if any of these conditions are hit.

File: src/core/qgsexpression.cpp [hunk 10]
```diff
@@ -2187,6 +2188,38 @@ static QVariant fcnMakePolygon( const QVariantList& values, const QgsExpressionC
...
 }
 
+static QVariant fcnMakeTriangle( const QVariantList& values, const QgsExpressionContext*, QgsExpression* parent )
+{
+  if ( values.count() < 3 )
+  {
+    return QVariant();
+  }
+
+  QgsTriangle *tr = new QgsTriangle();
```
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: switch these two to unique_ptrs for memory safety

File: tests/src/core/testqgsexpression.cpp [hunk 11]
```diff
@@ -765,6 +765,8 @@ class TestQgsExpression: public QObject
       QTest::newRow( "make_polygon" ) << "geom_to_wkt(make_polygon(geom_from_wkt('LINESTRING( 0 0, 0 1, 1 1, 1 0, 0 0 )')))" << false << QVariant( "Polygon ((0 0, 0 1, 1 1, 1 0, 0 0))" );
       QTest::newRow( "make_polygon rings" ) << "geom_to_wkt(make_polygon(geom_from_wkt('LINESTRING( 0 0, 0 1, 1 1, 1 0, 0 0 )'),geom_from_wkt('LINESTRING( 0.1 0.1, 0.1 0.2, 0.2 0.2, 0.2 0.1, 0.1 0.1 )'),geom_from_wkt('LINESTRING( 0.8 0.8, 0.8 0.9, 0.9 0.9, 0.9 0.8, 0.8 0.8 )')))" << false
       << QVariant( "Polygon ((0 0, 0 1, 1 1, 1 0, 0 0),(0.1 0.1, 0.1 0.2, 0.2 0.2, 0.2 0.1, 0.1 0.1),(0.8 0.8, 0.8 0.9, 0.9 0.9, 0.9 0.8, 0.8 0.8))" );
+      QTest::newRow( "make_triangle bad" ) << "make_triangle(make_point(2,4))" << false << QVariant();
+      QTest::newRow( "make_triangle" ) << "geom_to_wkt(make_triangle(make_point(0,0), make_point(5,5), make_point(0,10)))" << false << QVariant( "Triangle ((0 0, 5 5, 0 10, 0 0))" );
```
Comment (nyalldawson) [inline tests/src/core/testqgsexpression.cpp]: Can you also add tests which hit the three conditions checked in the loop? Ie null geometry/non-point geometries.

File: src/core/qgsexpression.cpp (hunk 8 above)
Comment (nyalldawson) [inline src/core/qgsexpression.cpp]: There's one part missing from this commit - you also need to add function help in https://github.com/qgis/QGIS/tree/master/resources/function_help/json

**Comment by nyalldawson** [comment]
@lbartoletti where's this at now? Ready to go?

@rouault can you re-review following all the updates?

**Review by rouault — APPROVED** [review]

**Comment by lbartoletti** [comment]
It's ok for me.
Thank you.

**Comment by nyalldawson** [comment]
Fantastic work - thank you!
